### PR TITLE
feat: @coinbase/wallet-sdk upgrade

### PR DESCRIPTION
**Pull request by nateReiners**
## Description

What changes are made in this PR? Is it a feature or a bug fix?
- This is just a draft to show what changes are needed in order to upgrade and get the ball rolling
- Upgrade `@coinbase/wallet-sdk` from v3 to v4
- v4 is almost ready and includes Coinbase smart wallets!

### Bundle size has been cut in half!
https://bundlephobia.com/package/@coinbase/wallet-sdk@4.0.0 

Before submitting this issue, please make sure you do the following.

- [ ] Read the [contributing guide](https://wagmi.sh/dev/contributing)
- [ ] Added documentation related to the changes made.
- [ ] Added or updated tests (and snapshots) related to the changes made.

### CODE REVIEW

File: packages/connectors/package.json [hunk 1]
```diff
@@ -45,7 +45,7 @@
     }
   },
   "dependencies": {
-    "@coinbase/wallet-sdk": "3.9.1",
+    "@coinbase/wallet-sdk": "4.0.0-beta.2",
```
Comment (nateReiners) [inline packages/connectors/package.json:48]: TODO: upgrade to v4 latest when published

**Comment by jxom** [comment]
What are the breaking changes in v4?

File: packages/connectors/src/coinbaseWallet.ts [hunk 2]
```diff
@@ -12,43 +10,14 @@ import {
...
-    reloadOnDisconnect?: boolean | undefined
-  }
->
-
 coinbaseWallet.type = 'coinbaseWallet' as const
-export function coinbaseWallet(parameters: CoinbaseWalletParameters) {
-  const reloadOnDisconnect = false
-
-  type Provider = CoinbaseWalletProvider
+export function coinbaseWallet(parameters: CoinbaseWalletSDKOptions) {
```
Comment (tmm) [inline packages/connectors/src/coinbaseWallet.ts:14]: Let's keep `CoinbaseWalletParameters` around, even if it just proxies `CoinbaseWalletSDKOptions`. Gives us more control over the connector's public API.

**Comment by nateReiners** [comment]
> What are the breaking changes in v4?

Hey @jxom , For most apps there won't be any necessary code changes other than a version bump. However we did remove a bunch of optional SDK options so apps that were passing in any of those will need to remove them. Details [here](https://github.com/coinbase/coinbase-wallet-sdk/blob/master/packages/wallet-sdk/docs/migration_guide.md)

**Comment by arein** [comment]
We need a `disableSmartWallet` option please - also can you please provide a bundle size comparison @nateReiners ? we have thousands of apps using web3modal and we serve your SDK

**Comment by glitch-txs** [comment]
hi @nateReiners we are also using `enableMobileWalletLink` in Web3Modal, do you know if this would cause some sort of error or conflict in the SDK? It would be a bit tricky for us since Wagmi is just a peer dependency and we cannot control which version devs are using.

**Comment by nateReiners** [comment]
> hi @nateReiners we are also using `enableMobileWalletLink` in Web3Modal, do you know if this would cause some sort of error or conflict in the SDK? It would be a bit tricky for us since Wagmi is just a peer dependency and we cannot control which version devs are using.

Hey @glitch-txs `enableMobileWalletLink` is gone, but the ability to connect via mobile is enabled by default in v4. When the SDK receives an `eth_requestAccounts` request, it will open a popup that allows the user to choose between connecting via smart wallet or Coinbase Wallet mobile. You can test out the UX for smart wallet or mobile wallet link here: https://www.smart-wallet.xyz/ Feedback welcome 

To remove the option to connect via mobile, dapps can pass in `smartWalletOnly: true` as an optional SDK option

**Comment by wilsoncusack** [comment]
> We need a `disableSmartWallet` option please - also can you please provide a bundle size comparison @nateReiners ? we have thousands of apps using web3modal and we serve your SDK

Hey Derek can you say more why the disable option is needed?

**Comment by glitch-txs** [comment]
@nateReiners thanks for the quick response! My question is if the SDK will throw an error if I pass it down the `enableMobileWalletLink` option in v4. In which case it would break Web3Modal if the dev is using an "old" version.

**Comment by nateReiners** [comment]
> also can you please provide a bundle size comparison

@arein Yep, just added it to the PR description. gzipped size was reduced by ~20 KB 🎉

**Comment by jxom** [comment]
> We need a disableSmartWallet option please - also can you please provide a bundle size comparison @nateReiners ? we have thousands of apps using web3modal and we serve your SDK

Curious why you'd want to disable smart wallets? Shouldn't Web3Modal be agnostic in this sense?

**Comment by nateReiners** [comment]
> @nateReiners thanks for the quick response! My question is if the SDK will throw an error if I pass it down the `enableMobileWalletLink` option in v4. In which case it would break Web3Modal if the dev is using an "old" version.

No the SDK wont throw an error. All deprecated options including `enableMobileWalletLink` are just ignored in v4. It would just be a Typescript issue.

**Comment by arein** [comment]
@jxom @wilsoncusack I can imagine some dapps not wanting this so I think it makes sense to allow them to configure. What rationale went into the option to exclusively use the smart wallet.

**Comment by glitch-txs** [comment]
Hi Coinbase team! Is this issue solved here? Also curious why the SDK is using the Solana package, is it meant to be multi-chain?
https://github.com/wevm/wagmi/issues/3836

**Comment by nateReiners** [comment]
Hey just wanted to drop an update here since I know this has been sitting for some time. We have been making some SDK interface tweaks and I wanted to wait until all of those decisions are solidified before cleaning up this PR. 

We are adding the ability to disable smart wallet as part of these tweaks via an 'eoaOnly' option cc: @arein

File: packages/connectors/src/coinbaseWallet.ts [hunk 3]
```diff
@@ -134,10 +111,10 @@ export function coinbaseWallet(parameters: CoinbaseWalletParameters) {
     },
     async getChainId() {
       const provider = await this.getProvider()
-      const chainId = await provider.request<number>({ method: 'eth_chainId' })
```
Comment (bangtoven) [inline packages/connectors/src/coinbaseWallet.ts:137]: do we need this change?

File: packages/connectors/src/coinbaseWallet.ts [hunk 4]
```diff
@@ -122,7 +100,6 @@ export function coinbaseWallet(parameters: CoinbaseWalletParameters) {
       }
 
       provider.disconnect()
-      provider.close()
```
Comment (bangtoven) [inline packages/connectors/src/coinbaseWallet.ts:125]: what if `provider`'s version is 3.x?

File: packages/connectors/src/coinbaseWallet.ts [hunk 5]
```diff
@@ -85,7 +63,7 @@ export function coinbaseWallet(parameters: CoinbaseWalletParameters) {
         }
 
         // Switch to chain if provided
-        let currentChainId = await this.getChainId()
+        let currentChainId: number = await this.getChainId()
```
Comment (bangtoven) [inline packages/connectors/src/coinbaseWallet.ts:66]: is this necessary?

File: packages/connectors/src/coinbaseWallet.ts [hunk 6]
```diff
@@ -1,268 +1,226 @@
...
-     * Fallback Ethereum Chain ID
-     * @default 1
-     */
-    chainId?: number | undefined
-    /**
-     * Whether or not to reload dapp automatically after disconnect.
-     * @default false
-     */
-    reloadOnDisconnect?: boolean | undefined
+  Mutable<Omit<ConstructorParameters<typeof CoinbaseWalletSDK>[0], "appChainIds">> & {
```
Comment (bangtoven) [inline packages/connectors/src/coinbaseWallet.ts:14]:
```suggestion
  Mutable<
    Omit<
      ConstructorParameters<typeof CoinbaseWalletSDK>[0],
      'appChainIds' // remove property and use `config.chains`
    >
  > & {
```

**Review by bangtoven — COMMENTED** [review]
nit: to minimize diff

File: packages/connectors/src/coinbaseWallet.ts (hunk 4 above)
Comment (nateReiners) [reply packages/connectors/src/coinbaseWallet.ts:125]: good catch, addressed in [886fb51](https://github.com/wevm/wagmi/pull/3784/commits/886fb51fd306b14e81ec5b1d3afeac2d00a42c15)

File: packages/connectors/src/coinbaseWallet.ts (hunk 3 above)
Comment (nateReiners) [reply packages/connectors/src/coinbaseWallet.ts:137]: no, removed

File: packages/connectors/src/coinbaseWallet.ts [hunk 7]
```diff
@@ -17,37 +14,22 @@ import {
...
-    chainId?: number | undefined
-    /**
-     * Whether or not to reload dapp automatically after disconnect.
-     * @default false
-     */
-    reloadOnDisconnect?: boolean | undefined
-  }
+      'appChainIds' // set via wagmi config
+    > & {
+      preference?: Preference
```
Comment (tmm) [inline packages/connectors/src/coinbaseWallet.ts:25]: Let's inline `Preference`

```suggestion
      preference?: Parameters<CoinbaseWalletSDK['makeWeb3Provider']>[0] | undefined
```

File: packages/connectors/src/coinbaseWallet.ts [hunk 8]
```diff
@@ -17,37 +14,22 @@ import {
   numberToHex,
 } from 'viem'
 
-// TODO(@3): Set `enableMobileWalletLink` to `true`
+type Preference = Parameters<CoinbaseWalletSDK['makeWeb3Provider']>[0]
+
```
Comment (tmm) [inline packages/connectors/src/coinbaseWallet.ts:18]:
```suggestion
```

**Review by tmm — COMMENTED** [review]
Docs will need updating too

https://wagmi.sh/react/api/connectors/coinbaseWallet

**Pull request closed without merging**